File: clients/libjabi/peripherals/i2c.cpp
#include <cstring>
#include <libjabi/byteorder.h>
#include <libjabi/interfaces/interface.h>

namespace jabi {

#include <jabi/peripherals.h>
#include <jabi/peripherals/i2c.h>

void Device::i2c_set_freq(I2CFreq preset, int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_I2C_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = I2C_SET_FREQ_ID,
        .payload_len = sizeof(i2c_set_freq_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<i2c_set_freq_req_t*>(req.payload);
    args->preset = static_cast<uint8_t>(preset);

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

void Device::i2c_write(int addr, std::vector<uint8_t> data, int idx) {
    if ((sizeof(i2c_write_j_req_t) + data.size()) > REQ_PAYLOAD_MAX_SIZE) {
        throw std::runtime_error("data too long");
    }

    iface_req_t req = {
        .periph_id   = PERIPH_I2C_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = I2C_WRITE_ID,
        .payload_len = static_cast<uint16_t>(sizeof(i2c_write_j_req_t) + data.size()),
        .payload     = {0},
    };

    auto args = reinterpret_cast<i2c_write_j_req_t*>(req.payload);
    args->addr = htole<uint16_t>(addr);
    args->data_len = htole<uint16_t>(data.size());
    memcpy(args->data, data.data(), data.size());

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != 0) {
        throw std::runtime_error("unexpected payload length");
    }
}

std::vector<uint8_t> Device::i2c_read(int addr, size_t len, int idx) {
    iface_req_t req = {
        .periph_id   = PERIPH_I2C_ID,
        .periph_idx  = static_cast<uint16_t>(idx),
        .periph_fn   = I2C_READ_ID,
        .payload_len = sizeof(i2c_read_j_req_t),
        .payload     = {0},
    };

    auto args = reinterpret_cast<i2c_read_j_req_t*>(req.payload);
    args->addr = htole<uint16_t>(addr);
    args->data_len = htole<uint16_t>(len);

    iface_resp_t resp = interface->send_request(req);
    if (resp.payload_len != len) {
        throw std::runtime_error("unexpected payload length");
    }

    auto ret = std::vector<uint8_t>(len, 0);
    memcpy(ret.data(), resp.payload, len);
    return ret;
}

};

File: clients/rust/src/peripherals/i2c.rs
use crate::interfaces::InterfaceRequest;
use crate::Error;
use deku::{DekuContainerWrite, DekuUpdate, DekuWrite};

pub enum I2CFreq {
    Standard = 0, // 100kHz
    Fast = 1,     // 400kHz
    FastPlus = 2, // 1MHz
    High = 3,     // 3.4MHz
    Ultra = 4,    // 5MHz
}

enum Func {
    SetFreq = 0,
    Write = 1,
    Read = 2,
    Transceive = 3,
}

fn gen_req(f: Func, idx: usize, payload: Vec<u8>) -> Result<InterfaceRequest, Error> {
    let mut r = InterfaceRequest {
        periph_id: crate::InstID::I2C as u16,
        periph_idx: idx as u16,
        periph_fn: f as u16,
        payload_len: 0,
        payload: payload,
    };
    r.update().map_or(Err(Error::PacketFormat), |_| Ok(r))
}

impl crate::Device {
    pub fn i2c_set_freq(&self, idx: usize, preset: I2CFreq) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct SetFreqRequest {
            preset: u8,
        }
        let req = SetFreqRequest {
            preset: preset as u8,
        };
        let resp = self.send(&gen_req(Func::SetFreq, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn i2c_write(&self, idx: usize, addr: u16, data: &[u8]) -> Result<(), Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct WriteRequest<'a> {
            addr: u16,
            data: &'a [u8],
        }
        let req = WriteRequest { addr, data };
        let resp = self.send(&gen_req(Func::Write, idx, req.to_bytes().unwrap())?)?;
        if resp.len() == 0 {
            Ok(())
        } else {
            Err(Error::PacketFormat)
        }
    }

    pub fn i2c_read(&self, idx: usize, addr: u16, len: usize) -> Result<Vec<u8>, Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct ReadRequest {
            addr: u16,
            data_len: u16,
        }
        let req = ReadRequest {
            addr,
            data_len: len as u16,
        };
        let resp = self.send(&gen_req(Func::Read, idx, req.to_bytes().unwrap())?)?;
        Ok(resp)
    }

    pub fn i2c_transceive(
        &self,
        idx: usize,
        addr: u16,
        data: &[u8],
        read_len: usize,
    ) -> Result<Vec<u8>, Error> {
        #[derive(DekuWrite)]
        #[deku(endian = "little")]
        struct TransceiveRequest<'a> {
            addr: u16,
            data_len: u16,
            data: &'a [u8],
        }
        let req = TransceiveRequest {
            addr,
            data_len: read_len as u16,
            data,
        };
        let resp = self.send(&gen_req(Func::Transceive, idx, req.to_bytes().unwrap())?)?;
        Ok(resp)
    }
}
